### PR TITLE
Fix register implement local

### PR DESCRIPTION
**Pull request by AlexMelov**
Register new user, return back token on login and set feature for delete user on backend only, also finished jest tests for the backend part

### CODE REVIEW

File: backend/src/controlers/mailer.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import { createTransport, Transporter } from 'nodemailer';
+import SMTPTransport from 'nodemailer/lib/smtp-transport';
+import { User } from '../models/user.interface';
+
+export function sendRegisterMail(user : User) : void
+{
+	//todo add multi-language support
+	//todo separate register message into other
+	const registerMessage : {} =
```
Comment (henryruhs) [inline backend/src/controlers/mailer.ts]: Define Message model here

File: backend/src/controlers/mailer.ts [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+	transport()
+		.sendMail(registerMessage)
+		.then(result => result)
+		.catch(error => ({ message: error }));
+}
+export function sendLoginMail(user : User) : void
+{
+	//todo separate message and add multi language
+	const loginMessage : {} =
```
Comment (henryruhs) [inline backend/src/controlers/mailer.ts]: Define Message model here

File: backend/src/controlers/mailer.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+			text: 'Your registration is done!'
+		};
+
+	transport()
+		.sendMail(loginMessage)
+		.then(result => result)
+		.catch(error => ({ message: error }));
+}
+
+//todo add environment variables
```
Comment (henryruhs) [inline backend/src/controlers/mailer.ts]: Use environment files here and remove credentials

File: backend/src/controlers/mailer.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+function transport() : Transporter<SMTPTransport.SentMessageInfo>
+{
+	return createTransport(
+	{
+		host: 'smtp.ethereal.email',
+		port: 587,
+		auth: {
```
Comment (henryruhs) [inline backend/src/controlers/mailer.ts]: { on new line

File: backend/src/controlers/delete.user.handler.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import { Request, Response } from 'express';
```
Comment (henryruhs) [inline backend/src/controlers/delete.user.handler.ts]: Rename file to users.ts

File: backend/src/controlers/login.handler.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import { Request, Response } from 'express';
```
Comment (henryruhs) [inline backend/src/controlers/login.handler.ts]: Rename to authentication.ts

File: backend/src/controlers/login.handler.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import { User } from '../models/user.interface';
+import { compare } from 'bcrypt';
+import pkg from 'jsonwebtoken';
```
Comment (henryruhs) [inline backend/src/controlers/login.handler.ts]: Please use `import * as jwt from 'jsonwebtoken';`

File: backend/src/controlers/login.handler.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+		.then(result =>
+		{
+			if(!result)
+			{
+				return response.status(401).json(
+				{
+					message: 'Authentication failed on user'
+				});
+			}
+			const { sign } = pkg;
```
Comment (henryruhs) [inline backend/src/controlers/login.handler.ts]: Try to import sign directly via import

File: backend/src/controlers/login.handler.ts [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+			const token : Token = sign(
```
Comment (henryruhs) [inline backend/src/controlers/login.handler.ts]: You will need to create the token more often - please move to token.helper.ts and write testing for it.

File: backend/src/controlers/login.handler.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+			response.status(200).json(
+			{
+				token
+			});
+		})
+		.catch(error =>
+		{
+			return response.status(401).json(
+			{
+				message: 'Authentication failed on entire', error
```
Comment (henryruhs) [inline backend/src/controlers/login.handler.ts]: Dont use text directly here... there should be a wording.ts with all the wording.

File: backend/src/controlers/mailer.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+		.catch(error => ({ message: error }));
+}
+export function sendLoginMail(user : User) : void
+{
+	//todo separate message and add multi language
+	const loginMessage : {} =
+		{
+			from: '"Sender Name" <theExpressApp@example.net>',
+			to: user.email,
+			subject: 'Registration',
```
Comment (henryruhs) [inline backend/src/controlers/mailer.ts]: Use from wording.ts

File: backend/src/controlers/register.handler.ts [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+			user.save()
+				.then(user =>
+				{
+					sendRegisterMail(user);
+					response.json(user);
+				})
+				.catch(error => response.status(403).json({ message: error }));
+		})
+		.catch(error => response.json({ message: error }));
```
Comment (henryruhs) [inline backend/src/controlers/register.handler.ts]: You are not sending http error code here

File: backend/src/controlers/mailer.ts [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+		from: '"Sender Name" <theExpressApp@example.net>',
+		to: user.email,
+		subject: 'Registration',
+		text: 'Your registration is done!'
+	};
+
+	transport()
+		.sendMail(registerMessage)
+		.then(result => result)
+		.catch(error => ({ message: error }));
```
Comment (henryruhs) [inline backend/src/controlers/mailer.ts:23]: You are not sending http error code here

File: backend/src/controlers/delete.user.handler.ts [hunk 14]
```diff
@@ -0,0 +1,10 @@
...
+import { Model } from 'mongoose';
+import { User } from '../models/user.interface';
+
+export function deleteUserHandler(request : Request, response : Response, userModel : Model<User>) : void
+{
+	userModel.deleteOne({ _id: request.params.userId })
+		.then(data => response.json(data))
+		.catch(error => response.json({ message: error }));
```
Comment (henryruhs) [inline backend/src/controlers/delete.user.handler.ts]: You are not sending http error code here

File: backend/src/controlers/login.handler.ts [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+				{ expiresIn: '1h' });
+
+			response.status(200).json(
+			{
+				token
+			});
+		})
+		.catch(error =>
+		{
+			return response.status(401).json(
```
Comment (henryruhs) [inline backend/src/controlers/login.handler.ts]: Return is not needed... you can write this shorter

File: backend/src/controlers/todos.ts [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import { Request, Response } from 'express';
+import { HydratedDocument, Model } from 'mongoose';
+import { DirtyTodo, Todo } from '../models/todo';
+
+export function getAllTodos(request : Request, response : Response, todoModel : Model<Todo>) : void
+{
+	todoModel.find()
+		.then(data => response.json(mapData(data)))
+		.catch(error => response.json({ message: error }));
```
Comment (henryruhs) [inline backend/src/controlers/todos.ts]: You are not sending http error code here

File: backend/src/controlers/todos.ts [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export function postTodo(request : Request, response : Response, todoModel : Model<Todo>) : void
+{
+	const todo : HydratedDocument<Todo> = new todoModel(request.body);
+
+	todo.save()
+		.then(data => response.json(data))
+		.catch(error => response.json({ message: error }));
```
Comment (henryruhs) [inline backend/src/controlers/todos.ts]: You are not sending http error code here

File: backend/src/controlers/todos.ts [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+export function deleteOneTodo(request : Request, response : Response, todoModel : Model<Todo>) : void
+{
+	todoModel.deleteOne({ _id: request.params.todoId })
+		.then(data => response.json(data))
+		.catch(error => response.json({ message: error }));
```
Comment (henryruhs) [inline backend/src/controlers/todos.ts]: You are not sending http error code here

File: backend/src/controller.ts [hunk 19]
```diff
@@ -1,33 +1,38 @@
...
+	registerUser(request, response, userModel);
 };
 
-function mapData(data : DirtyTodo[])
+export const loginHandler : Handler = (request : Request, response : Response) : void =>
 {
-	return data.map(item =>({ id: item._id, title: item.title }));
-}
+	loginUserHandler(request, response, userModel);
+};
```
Comment (henryruhs) [inline backend/src/controller.ts:33]: Remove empty line

File: backend/src/environments/environment.dev.ts [hunk 20]
```diff
@@ -12,7 +12,10 @@ export const environment : Environment =
 	apiRoutes:
 	{
 		todos: '/api/todos',
-		todosWithId: '/api/todos/:todoId'
+		todosWithId: '/api/todos/:todoId',
+		register: '/register',
+		login: '/login',
+		apiWithId: '/register/:userId'
```
Comment (henryruhs) [inline backend/src/environments/environment.dev.ts]: register: '/register' without userId ... please send erverything via POST

File: backend/src/models/todo.ts [hunk 21]
```diff
@@ -8,3 +8,8 @@ export interface DirtyTodo
 	title : {};
 	_id : {};
 }
+export interface TestTodo
```
Comment (henryruhs) [inline backend/src/models/todo.ts:12]: Newline before interface

File: backend/src/server.spec.ts [hunk 22]
```diff
@@ -38,7 +42,31 @@ describe('Server', () =>
...
+
+		expect(body.forEach(todo =>
+		{
+			expect(todo.title).toBeTruthy();
+		}));
+		await supertest(server).delete('/todos/' + todoArray[0].id).then(user =>
+		{
+			expect(user.statusCode).toBe(200);
+			expect(user.body).toEqual(
+				{
```
Comment (henryruhs) [inline backend/src/server.spec.ts]: Wrong indent

File: backend/src/server.spec.ts [hunk 23]
```diff
@@ -38,7 +42,31 @@ describe('Server', () =>
 		expect(body.title).toContain('Todo from Jest!');
 	});
 
-	it('should DELETE todo', async() =>
+	it('Should GET last added todo from list and delete it', async() =>
+	{
+		const response : Response = await supertest(server).get('/todos');
+		const body : TestTodo[] = await response.body;
+
+		expect(response.statusCode).toBe(200);
```
Comment (henryruhs) [inline backend/src/server.spec.ts]: expect on the end after const

File: backend/src/server.spec.ts [hunk 24]
```diff
@@ -60,5 +88,107 @@ describe('Server', () =>
...
+			password : '123456'
+		});
+		const body : User = await response.body;
+
+		expect(response.statusCode).toBe(200);
+		expect(response.headers).toBeDefined();
+		expect(body.email).toContain('jest.email@mail.com');
+		expect(body.password).not.toHaveLength(0);
+
+		const { _id } = response.body;
```
Comment (henryruhs) [inline backend/src/server.spec.ts]: const always at the top

File: backend/src/server.spec.ts [hunk 25]
```diff
@@ -60,5 +88,107 @@ describe('Server', () =>
...
+			password : '123456'
+		});
+		const body : User = await response.body;
+
+		expect(response.statusCode).toBe(200);
+		expect(response.headers).toBeDefined();
+		expect(body.email).toContain('jest_repeat.email@mail.com');
+		expect(body.password).not.toHaveLength(0);
+
+		const repeatedUserResponse : Response = await supertest(server)
```
Comment (henryruhs) [inline backend/src/server.spec.ts]: const always at the top

File: backend/src/server.spec.ts [hunk 26]
```diff
@@ -2,9 +2,13 @@ import supertest, { Response } from 'supertest';
 import mongoose from 'mongoose';
```
Comment (henryruhs) [inline backend/src/server.spec.ts:2]: You are not doing a server.spec.ts ... you have to create one spec file for each controller instead like todos.spec.ts and register.spec.ts

File: backend/src/controlers/register.handler.ts [hunk 27]
```diff
@@ -0,0 +1,29 @@
+import { Request, Response } from 'express';
```
Comment (henryruhs) [inline backend/src/controlers/register.handler.ts]: Rename to register.ts

File: frontend/cypress/integration/todo.spec.ts [hunk 28]
```diff
@@ -31,6 +31,7 @@ describe('todo', () =>
 
 	it('removing todo from list', () =>
 	{
+		cy.reload();
```
Comment (henryruhs) [inline frontend/cypress/integration/todo.spec.ts]: There should be no reload

File: frontend/src/app/app-routing.module.ts [hunk 29]
```diff
@@ -0,0 +1,19 @@
...
+		RouterModule.forRoot(routes)
+	],
+	exports:
+	[
+		RouterModule
+	]
+})
+export class AppRoutingModule
+{
+
```
Comment (henryruhs) [inline frontend/src/app/app-routing.module.ts]: remove empty line

File: frontend/src/app/authentication/authorisation-interceptor.ts [hunk 30]
```diff
@@ -0,0 +1,26 @@
...
+{
+	providedIn:'root'
+})
+export class AuthorisationInterceptor implements HttpInterceptor
+{
+	constructor(private loginComponent : LoginComponent)
+	{
+	}
+
+	intercept(request : HttpRequest<any>, next : HttpHandler) : Observable<HttpEvent<any>>
```
Comment (henryruhs) [inline frontend/src/app/authentication/authorisation-interceptor.ts:16]: you know that any is not allowed... CI should fail here

File: frontend/src/app/authentication/login/login.component.scss [hunk 31]
```diff
@@ -0,0 +1,14 @@
+.login-form
+{
+	align-content: center;
+	display: flex;
+	flex-direction: column;
+	justify-content: center;
+	width: 25em;
+
+	.signUp-button
```
Comment (henryruhs) [inline frontend/src/app/authentication/login/login.component.scss]: no camelcase in css

File: frontend/src/app/authentication/login/login.component.scss [hunk 32]
```diff
@@ -0,0 +1,14 @@
+.login-form
```
Comment (henryruhs) [inline frontend/src/app/authentication/login/login.component.scss]: form-login

File: frontend/src/app/authentication/login/login.component.scss (hunk 31 above)
Comment (henryruhs) [inline frontend/src/app/authentication/login/login.component.scss]: button-login

File: frontend/src/app/authentication/login/login.component.ts [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+
+@Component(
+{
+	selector: 'app-login',
+	templateUrl: './login.component.html',
+	styleUrls: [ './login.component.scss' ]
+})
+export class LoginComponent
+{
+	unmask : boolean = true;
```
Comment (henryruhs) [inline frontend/src/app/authentication/login/login.component.ts]: unmask should be false by default

File: frontend/src/app/authentication/register/register.component.scss [hunk 34]
```diff
@@ -0,0 +1,14 @@
+.register-form
```
Comment (henryruhs) [inline frontend/src/app/authentication/register/register.component.scss]: form-register

File: frontend/src/app/authentication/register/register.component.scss [hunk 35]
```diff
@@ -0,0 +1,14 @@
...
+{
+	align-content: center;
+	display: flex;
+	flex-direction: column;
+	justify-content: center;
+	width: 25em;
+
+	.signUp-button
```
Comment (henryruhs) [inline frontend/src/app/authentication/register/register.component.scss]: button-register

File: frontend/src/app/authentication/register/register.component.ts [hunk 36]
```diff
@@ -0,0 +1,44 @@
...
+@Component(
+{
+	selector: 'app-register',
+	templateUrl: './register.component.html',
+	styleUrls: [ './register.component.scss' ]
+}
+)
+export class RegisterComponent
+{
+	unmask : boolean = true;
```
Comment (henryruhs) [inline frontend/src/app/authentication/register/register.component.ts]: Should be false by default

File: frontend/src/app/todo/form/form.component.spec.ts [hunk 37]
```diff
@@ -12,7 +13,7 @@ describe('ListComponent', () =>
 	{
 		await TestBed.configureTestingModule(
 			{
```
Comment (henryruhs) [inline frontend/src/app/todo/form/form.component.spec.ts]: Wrong indent

File: frontend/src/app/todo/list/list.component.spec.ts [hunk 38]
```diff
@@ -3,7 +3,7 @@ import { HttpClientTestingModule } from '@angular/common/http/testing';
 import { TodoService } from '../todo.service';
 import { Todo } from '../todo.intercace';
 
-describe('ListComponent', () =>
+describe.skip('ListComponent', () =>
```
Comment (henryruhs) [inline frontend/src/app/todo/list/list.component.spec.ts]: Remove skip - no PR with skipped testing

File: frontend/src/app/todo/todo.component.spec.ts [hunk 39]
```diff
@@ -3,7 +3,7 @@ import { TestBed } from '@angular/core/testing';
 import { HttpClientTestingModule } from '@angular/common/http/testing';
 import { TodoService } from './todo.service';
 
-describe('ListComponent', () =>
+describe.skip('ListComponent', () =>
```
Comment (henryruhs) [inline frontend/src/app/todo/todo.component.spec.ts]: Why that? Dont break the develop branch

File: frontend/src/app/todo/todo.service.ts [hunk 40]
```diff
@@ -1,32 +1,15 @@
...
 
-@Injectable({
+@Injectable(
+{
 	providedIn: 'root'
 })
-export class TodoService
+@ApiUrl(environment.apiUrl)
+@ApiRoute(environment.apiRoutes.todos)
+export class TodoService extends CrudService<Todo, Todo, Todo, Todo, Todo, Todo[], Todo, Todo, Todo, Todo, void>
```
Comment (henryruhs) [inline frontend/src/app/todo/todo.service.ts]: You did not merge develop branch into that one - this is false commit and we are using crud service custom !!!

**Review by henryruhs — CHANGES_REQUESTED** [review]

File: backend/src/controlers/login.handler.ts (hunk 7 above)
Comment (AlexMelov) [reply backend/src/controlers/login.handler.ts]: It's not working that way I've tried already last week

File: backend/src/controlers/login.handler.ts (hunk 8 above)
Comment (AlexMelov) [reply backend/src/controlers/login.handler.ts]: Not working, we were considering this with Dejan last week. This way was suggested from the terminal

File: frontend/src/app/authentication/authorisation-interceptor.ts (hunk 30 above)
Comment (AlexMelov) [reply frontend/src/app/authentication/authorisation-interceptor.ts:16]: I know but I don't know what to set there, I get like this from the documentation

File: backend/src/controlers/login.handler.ts (hunk 10 above)
Comment (AlexMelov) [reply backend/src/controlers/login.handler.ts]: this wording.ts what type of file should be, I set inside messageContext but should be something else inside like error message return string functions?

File: backend/src/environments/environment.dev.ts (hunk 20 above)
Comment (AlexMelov) [reply backend/src/environments/environment.dev.ts]: I'm not clear on this one

**Pull request closed without merging**